### PR TITLE
feat(cluster): Implement initial support for "hardware mappings"

### LINKED ISSUE #886: Support for "hardware mapping" as dedicated Terraform resource & data source
First of all: Happy new year and thanks for this provider, this is definitely the best to automate everything for a home lab 😄

I recently set up a new [OPNsense](https://opnsense.org) router device ([Protectli VP2420](https://protectli.com/product/vp2420)) as an additional Proxmox node. The whole VM is automatically created via [Packer](https://www.packer.io) which results in a VM template in the end. The actual VM is managed with Terraform, using this provider. Since the [NIC](https://en.wikipedia.org/wiki/Network_interface_controller) that acts as the [WAN](https://en.wikipedia.org/wiki/Wide_area_network) in _OPNSense_ should only be used by this VM, and also to improve the performance, I pass it through as PCI device.

After revisiting my code after the holidays I noticed that [it is also possible to use a mapped resource](https://github.com/bpg/terraform-provider-proxmox/pull/500), but that there is no dedicated `proxmox_virtual_environment_cluster_hardware_mapping` resource but that this step must still be done manually (or automated through other ways that interact with the Proxmox API). Even though [the Proxmox API endpoint for hardware mappings](https://pve.proxmox.com/pve-docs/api-viewer/#/cluster/mapping/pci) is quite confusing I guess it might be nice if this provider could allow to manage this resource too.

I know that this would add another point in the "Known Issues" documentation since it is only possible to map a PCI/USB device using the `root` PAM account, but I guess this is still better than having to use my mouse to click it… because [IaC](https://en.wikipedia.org/wiki/Infrastructure_as_code) everything 😀

An alternative solution I've considered was to add this PCI passthrough mapping to my _Packer_ configuration, but that is not an ideal solution as it would bind it to the template which should rather act as a "one-time VM template" that gets deleted after (fully) cloning the VM with Terraform. This is also only a solution for me and does not allow to manage hardware mappings with Terraform in a general way.

Looking forward to hear your opinion! I'm also open to contribute this feature if you're busy, or support in any way to test it if you'd implement it yourself.

### PR DESCRIPTION
**Pull request by svengreb**
### Summary

Right now [it is already possible to use a mapped resource][1], but there is no dedicated `proxmox_virtual_environment_cluster_hardware_mapping` resource but this step must still be done manually (or automated through other ways that interact with the Proxmox API).

This pull request implements support for "hardware mapping" resources and data sources for the, currently, available bus types PCI and USB, based on [the Proxmox VE API documentations][2].

There are some "specialties" in these resources and data sources:

1. The Proxmox VE API uses the `description` attribute, but this implementations names it `comment` since this naming is generally across the Proxmox VE web UI and API documentations. This still follows [the Terraform "best practices"][3] as it improves the user experience by matching the field name to the naming used in the human-facing interfaces.
2. Like in point 1, the name of the attribute of "node checks diagnostics" for USB hardware mappings is `errors` in the Proxmox VE API while it is `checks` for hardware mappings of type PCI. The second naming pattern is also generally used across the Proxmox VE web UI and API documentations, including the `check_node` attribute that is also implemented in the `proxmox_virtual_environment_hardware_mappings` data source. Therefore, this implementation named both attributes "checks" which still follows [the Terraform "best practices"][3] as it improves the user experience by matching the field name to the naming used in the human-facing interfaces.
3. This implementation comes with the "unique" feature of allowing comments (named `descriptions` by the Proxmox VE API) for an entry in a device map which is not possible through the web UI at all but only adding a comment for the whole mapping entry instead.

Note that this implementation also adds another point in the "Known Issues" documentation since it is only possible to map a PCI/USB device using the `root` PAM account, but this is still better than having to manually configure it through the web UI or by interacting with the Proxmox VE API on other ways.

### Contributor's Note

- [x] I have added / updated documentation in `/docs` for any user-facing features or additions.
- [x] I have added / updated acceptance tests in `/fwprovider/tests` for any new or updated resources / data sources.
- [x] I have ran `make example` to verify that the change works as expected.

### Proof of Work

<p align="center">
 <figure>
    
    <figcaption>Overview of all cluster hardware mappings with the example configurations applied.</figcaption>
  </figure> 
</p>

### Community Note

- Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Related to #500
Closes #886

[1]: https://github.com/bpg/terraform-provider-proxmox/pull/500
[2]: https://pve.proxmox.com/pve-docs/api-viewer/#/cluster/mapping/pci
[3]: https://developer.hashicorp.com/terraform/plugin/best-practices/hashicorp-provider-design-principles#resource-and-attribute-schema-should-closely-match-the-underlying-api

### CODE REVIEW

**Comment by bpg** [comment]
Awesome! @svengreb thanks so much for contributing this!
That's a lot of code, will take a bit of time to review :)

File: fwprovider/resource_hardware_mapping_pci.go [hunk 1]
```diff
@@ -0,0 +1,348 @@
+/*
+	This Source Code Form is subject to the terms of the Mozilla Public
+	License, v. 2.0. If a copy of the MPL was not distributed with this
+	file, You can obtain one at https://mozilla.org/MPL/2.0/.
+*/
+
+//nolint:nolintlint,gofumpt,wsl // wsl linter is random-linter-loop-broken and only causes code to become unreadable.
```
Comment (bpg-dev) [inline fwprovider/resource_hardware_mapping_pci.go]: This line can be removed (here and in _usb.go file). The only linter warning it suppresses is lines 61-62, which can be easily fixed by adding an empty line on 60.

**Review by bpg — COMMENTED** [review]
Would it be possible to rearrange the code a bit, moving everything under e.g. `hardware_mapping` (or `hw_mappings`, or whatever is ) package?
Screenshot 2024-04-15 at 6 08 57 PM

Then we can shorten names of the constants / types, etc.
I'll refactor other resources / datasources in the same way for consistency

**Comment by svengreb** [comment]
> Awesome! @svengreb thanks so much for contributing this! That's a lot of code, will take a bit of time to review :)

Sure, take your time!
It's indeed a lot, but I guess this is always the ase for adding a whole new feature from th Proxmox VE API.

**Comment by svengreb** [comment]
> Would it be possible to rearrange the code a bit, moving everything under e.g. `hardware_mapping` (or `hw_mappings`, or whatever is ) package?

Sure, I only adapted the existing layout, but I'm also a fan of sub-packages in Go projects instead of completely cluttered folders. This also always simplifies the code because the need to prefix everything is gone.

I'd leave the name of the package more explicit (`hardware_mapping`, singular like in the Proxmox VE API) and use `HWM` (maybe even `HM`) as prefix inside the code, but only when it is necessary.

**Comment by svengreb** [comment]
@bpg The layout refactoring is done. The package names `hardware_mapping` could not be used because of the underscore, so I simply named it `hardwaremapping` (could maybe shortened to `hwmapping`, but being more explicit also helps to understand code).

**Comment by bpg** [comment]
> @bpg The layout refactoring is done. The package names `hardware_mapping` could not be used because of the underscore, so I simply named it `hardwaremapping` (could maybe shortened to `hwmapping`, but being more explicit also helps to understand code).

Thanks @svengreb! That looks good.
Would you mind to run `make lint` one more time to fix the formatting, and then to [sing-off your commits](https://github.com/bpg/terraform-provider-proxmox/pull/1213/checks?check_run_id=23912763506)?

**Comment by svengreb** [comment]
Sure, but it's very strange: when linting uncommitted changes the `wsl` and `gofumpt` linters both blocked themself, but running it afterwards the `--fix` option is able to resolve the problem without any of them complaining.
Anyway, fixed the last finding and signed-off all commits (I thought the first commit will be used for squashing so I only signed this one off).

**Review by bpg — APPROVED** [review]
LGTM! 🚀

**Pull request merged**